src/halmos/mapper.py

```diff
@@ -0,0 +1,173 @@
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Type
+
+
+@dataclass
+class AstNode:
+    node_type: str
+    id: int
+    name: str
+    address: str
+    visibility: str
+
+
+@dataclass
+class ContractMappingInfo:
+    contract_name: str
+    bytecode: str
+    nodes: List[AstNode]
+
+
+class SingletonMeta(type):
+    _instances: Dict[Type, Any] = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super().__call__(*args, **kwargs)
+
+        return cls._instances[cls]
+
+
+class Mapper(metaclass=SingletonMeta):
+    _PARSING_IGNORED_NODE_TYPES = [
+        "StructDefinition",
+        "EnumDefinition",
+        "PragmaDirective",
+        "ImportDirective",
+        "Block",
+    ]
+
+    def __init__(self):
+        self._contracts: Dict[str, ContractMappingInfo] = {}
+
+    def add_contract_mapping_info(
+        self, contract_name: str, bytecode: str, nodes: List[AstNode]
+    ):
+        if contract_name in self._contracts:
+            raise ValueError(f"Contract {contract_name} already exists")
+
+        self._contracts[contract_name] = ContractMappingInfo(
+            contract_name, bytecode, nodes
+        )
+
+    def get_contract_mapping_info_by_name(
+        self, contract_name: str
+    ) -> Optional[ContractMappingInfo]:
+        return self._contracts.get(contract_name, None)
+
+    def get_contarct_mapping_info_by_bytecode(
+        self, bytecode: str
+    ) -> Optional[ContractMappingInfo]:
+        for contract_mapping_info in self._contracts.values():
+            if contract_mapping_info.bytecode.endswith(bytecode):
```

Comment:
I don't think this works for contracts with immutable variables. It's fine for now, can leave it as a TODO

Reply:
quick confirm, adding this to BadElections.sol:

```solidity
    address immutable public owner;

    constructor() {
        owner = msg.sender;
    }
```

results:

```
# before
STATICCALL BadElections::Concat(votesFor, p_proposalId_uint256()) [static]

# after
STATICCALL 0xaaaa0002::Concat(BadElections.votesFor, p_proposalId_uint256()) [static]
```

Reply:
@karmacoma-eth 
Thank you for feedback!
I added TODO comment at commit 0611dd4